File: 32_octal_to_decimal/32_octal_to_decimal.dart
import 'dart:io';
import 'dart:math';

// program to convert coctal number to equivalent decimal number

int octalToDecimal(int octal) {
  int power = 0;
  int decimalSum = 0;
  String octalString = octal.toString();
  for (int i = octalString.length - 1; i <= 0; i--) {
    decimalSum =
        decimalSum + (int.parse(octalString[i]) * pow(8, power).toInt());
    power = power + 1;
  }
  return decimalSum;
}

bool isValidOctal(int num) {
  String numStr = num.toString();
  for (int i = 0; i < numStr.length; i++) {
    int digit = int.parse(numStr[i]);

    if (digit < 0 || digit > 7) {
      return false;
    }
  }
  return true;
}

void main() {
  try {
    while (true) {
      stdout.write("Enter a positive integer value : ");
      int num1 = int.parse(stdin.readLineSync()!);

      if (num1 < 0) {
        print("Enter the value in decimal format");
      } else if (isValidOctal(num1)) {
        print("Invalid octal number");
      } else {
        int result = octalToDecimal(num1);
        print("Octal $num1 = $result Decimal");
      }
      stdout.write("Do you want to continue the program (y/n) : ");
      String choice = stdin.readLineSync()!.toLowerCase();
      if (choice == 'n') {
        print("Program finished!");
        break;
      }
    }
  } on FormatException catch (e) {
    print("Invalid input : $e");
  } on UnsupportedError catch (e) {
    print("Divide by zero exception : $e");
  } on Exception catch (e) {
    print("Exception caught : $e");
  }
}

File: 32_octal_to_decimal/32_octal_to_decimal.py
# Python program to convert octal to decimal

# method 1 - brut-force and iteration
try:
    n=int(input("Enter the Octal number :"))
    st=str(n)[::-1]
    sum=0
    for i in range(len(st)):
        a=int(st[i])*(8**(i))
        sum=sum+a
    print(sum)
except:
    print("Enter a valid integer value.")

# method 2 - using recursion
try:
    n=int(input("Enter the Octal number :"))
    sum=0
    stt=str(n)[::-1]
    def oct_to_dec(p):
        if p>=len(stt):
            return print(end="")
        a=int(stt[p])*(8**(p))
        global sum
        sum=sum+a
        oct_to_dec(p+1)
    oct_to_dec(0)
    print(sum)
except:
    print("Enter a valid integer value.")

# method 3 - using built-in function
try:
    n=int(input("Enter the octal number :"))
    octl=int(str(n),8)
    print(octl)
except:
    print("Enter a valid integer value.")

# method 4 - using module based approach

def octal_to_decimal(octal):
    pow = 0
    decimal_sum = 0
    for i in str(octal)[::-1]:
        decimal_sum = decimal_sum + (int(i) * (8 ** pow))
        pow+=1
    return decimal_sum

def is_valid_octal(num):
    for i in str(num):
        if i not in [0,1,2,3,4,5,6,7]:
            return False
    return True

try:
    while(True):
        num = int(input("Enter positive integer value : "))
        if (num < 0):
            print("Please enter a valid octal number.")
        elif (is_valid_octal(num)):
            print("Invalid octal number")
        else:
            result = octal_to_decimal(num)
            print(f"Octal {num} = {result} Decimal")

        choice = input("Do you want to continue the program (y/n) : ")
        if(choice == "n"):
            print("Program finished!")
            break

except ValueError as e:
    print(f"Invalid Input : {e}")

except ZeroDivisionError as e:
    print(f"Zero division exception : {e}")

except Exception as e:
    print(f"Exception caught : {e}")


